controller/worker/domain_migration/domain_migration.go

```diff
@@ -0,0 +1,306 @@
+	}, events)
+	if err != nil {
+		return err
+	}
+	defer stream.Close()
+
+	timeout := time.After(time.Minute)
+	for {
+		select {
+		case event := <-events:
+			var data *ct.DeploymentEvent
+			if err := json.Unmarshal(event.Data, &data); err != nil {
+				return err
+			}
+			if data.Status == "complete" {
+				return nil
+			}
+			if data.Status == "failed" {
+				return errors.New(data.Error)
+			}
+		case <-timeout:
+			return errors.New("timed out waiting for deployment")
+		}
+	}
+}
+
+func (c *context) maybeDeployController(oldCert, cert *tlscert.Cert, oldDomain, newDomain string) error {
+	const appName = "controller"
+	if err := c.waitForDeployment(appName); err != nil {
+		return err
+	}
+	release, err := c.client.GetAppRelease(appName)
+	if err != nil {
+		return err
+	}
+	if release.Env["DEFAULT_ROUTE_DOMAIN"] != oldDomain {
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["DEFAULT_ROUTE_DOMAIN"] = newDomain
+	release.Env["CA_CERT"] = cert.CACert
+	if err := c.client.CreateRelease(release); err != nil {
+		return err
+	}
+	return c.client.DeployAppRelease(appName, release.ID)
+}
+
+func (c *context) maybeDeployRouter(oldCert, cert *tlscert.Cert, oldDomain, newDomain string) error {
+	const appName = "router"
+	if err := c.waitForDeployment(appName); err != nil {
+		return err
+	}
+	release, err := c.client.GetAppRelease(appName)
+	if err != nil {
+		return err
+	}
+	if release.Env["TLSCERT"] != oldCert.Cert {
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["TLSCERT"] = cert.Cert
+	release.Env["TLSKEY"] = cert.PrivateKey
+	if err := c.client.CreateRelease(release); err != nil {
+		return err
+	}
+	return c.client.DeployAppRelease(appName, release.ID)
+}
+
+func (c *context) maybeDeployDashboard(oldCert, cert *tlscert.Cert, oldDomain, newDomain string) error {
+	const appName = "dashboard"
+	if err := c.waitForDeployment(appName); err != nil {
+		return err
+	}
+	release, err := c.client.GetAppRelease(appName)
+	if err != nil {
+		return err
+	}
+	if release.Env["DEFAULT_ROUTE_DOMAIN"] != oldDomain {
+		return nil
+	}
+	release = dupRelease(release)
+	release.Env["CA_CERT"] = cert.CACert
+	release.Env["DEFAULT_ROUTE_DOMAIN"] = newDomain
+	release.Env["CONTROLLER_DOMAIN"] = fmt.Sprintf("controller.%s", newDomain)
+	release.Env["URL"] = fmt.Sprintf("dashboard.%s", newDomain)
+	if err := c.client.CreateRelease(release); err != nil {
+		return err
+	}
+	return c.client.DeployAppRelease(appName, release.ID)
+}
+
+func (c *context) createMissingRoutes(oldCert, cert *tlscert.Cert, oldDomain, newDomain string) error {
+	apps, err := c.client.AppList()
+	if err != nil {
+		return err
+	}
+	errChan := make(chan error)
+	createMissingRoutes := func(app *ct.App) {
+		errChan <- c.appCreateMissingRoutes(app.ID, oldCert, cert, oldDomain, newDomain)
+	}
+	for _, app := range apps {
+		go createMissingRoutes(app)
+	}
+	for range apps {
+		if err := <-errChan; err != nil {
+			return err
```

Comment:
We can/should close `errChan` after this loop, but something like `for err := range errChan` will cause a deadlock.